# 7. 関数

プログラムが長くなると，部品を作って何度も再利用したくなります．

そのようなプログラムの部品のことを関数と呼びます. Pythonでは， def で定義します．

基本的な構文は以下の通りです．

```python
def 関数名(引数）:

　　関数本体
  
　　return 返値
```
関数は入力を出力に変換します．

   入力  => 関数   => 出力 
 
入力を「引数（ひきすう）」，出力を「返値（かえりち）」と呼びます．return文はなくても大丈夫です．その場合には返値はありません（Noneという特別なオブジェクトが返されます）．


● 例題７－１

整数nを引数として受け取り，print “#”; を n回繰り返し，長さnの#でできた棒を表示し，その後改行をする関数を作りなさい．

例えば，10 を渡すと

##########

が表示され，25を渡すと，

#########################

が表示されるような関数を作る，ということです．


In [ ]:
def sub(n):
    for i in range(n):
        print('#',end='')

n = int(input('n='))
sub(n)

● 例題７－２

２つの数字を引数とし，その和を返値として返す関数を記述し，その関数を用いて 3と5の和を計算せよ．また，100と200の和も計算せよ．

複数の引数はカンマで区切って関数に渡します．

関数の値は返値になります．


In [ ]:
def computeSum(x,y):
      return x+y 
    
print( computeSum(3,5) )
print( computeSum(100,200) )

● 例題７－３

フィボナッチ数 $F_n$ は

$
F_1 =1  \\
F_2 =1  \\
F_n= F_{n-1} +F_{n-2}
$

と再帰的に定義される数列です．$F_5$ を計算してください．

これは関数 F の中で同じ関数 F を呼び出す再帰的なプログラムで記述します．

In [ ]:
def F(n):
    if n==1 or n==2:
        return 1
    else:
        return F(n-1)+F(n-2)
print(F(5))

● 問題 7-1

3つの整数を引数として渡すと，その積を返す関数を記述せよ．その関数を用いて，$2*5*7$ と $123*456*234$ を計算せよ．

In [ ]:
def seki(a, b, c):
    s = a * b * c
    return s

d = 2
e = 5
f = 7
un = seki(d, e, f)
print(un)

g = 123
h = 456
i = 234
un2 = seki(g, h, i)
print(un2)

In [ ]:
def seki(a, b, c):
    s = a * b * c
    return s

d = 2
e = 5
f = 7
print(seki(d, e, f))

●問題 7-2

問題7-1の関数を，入力した3つの数を引数として3回呼び出しなさい．例えば8,13,4と入力したら，

########

#############

####

と，棒グラフのような表示をする，ということです．

In [ ]:
def sub(a, b, c):
    for i in range(a):
        print('#',end='')
    print()
    print()
    for i in range(b):
        print('#',end='')
    print()
    print()
    for i in range(c):
        print('#',end='')

d = int(input("a入力"))
e = int(input("b入力"))
f = int(input("c入力"))

sub(d, e, f)

In [ ]:
def sub(a, b, c):
    for i in range(a):
        print('#',end='')
    print()
    print()
    for i in range(b):
        print('#',end='')
    print()
    print()
    for i in range(c):
        print('#',end='')

d, e, f = map(int, input().split())
sub(d, e, f)

●問題 7-3
問題7-1の関数を使って，#でできた三角形を表示するプログラムを作りなさい．
例えば，このようなものを表示するプログラムです．

#
##
###
####
#####
######

In [ ]:
def tri(n):
    for i in range(1, n + 1):
        print(" " * ( n - i ) + "#" * ( 2 * i - 1 ) )

x = int(input())
tri(x)

●問題7-4


数aを渡すと，aが正なら1を，負なら -1を，0ならば0を返す関数を作りなさい．

In [ ]:
def judge(a):
    if a > 0:
        return 1
    elif a == 0:
        return 0
    else:
        return -1

a = int(input("aを入力"))
print(judge(a))

●問題7-5

数aを受け取り，数bをキーボードから入力し，a+bの値を返す関数を作りなさい．

次に，この関数を使って，数を10個入力してその合計を表示するプログラムを作りなさい．

In [ ]:
def tasu(a):
    b = int(input("b入力"))
    return a + b

print(tasu(1))

In [ ]:
def tasu(x):
    ten = 0
    for i in range(10):
        a = int(input(f"a[{i + 1}]入力"))
        ten = ten + a
    return ten

print(tasu(1))

●問題7-6

$n$ の階乗 $n!$ は

$
1! =1 \\
n!= n \times (n-1)! 
$

と再帰的に定義されます．$n!$ を再帰的に計算する関数 Factorial(n) を記述し，$5!$ を計算しなさい．


In [ ]:
def kaijo(n):
    if n == 1:
        return 1
    elif n >= 2:
        return kaijo(n - 1) * n

n = int(input("階乗を入力"))
print(kaijo(n))

# 8. 文字列の操作

文字列についてはすでに ' ' もしくは "  "で挟むと生成できることを知っていると思います．

文字列は s ='Hello'は H e l l o が順番に箱に入ってしまわれています．前から順番に 0,1,2, ... と番号付け（添え字とかインデックスといいます）がされています．

文字 'H'を取り出すには，s[0] とします．

文字 'o' を取り出すには s[4] とします．

文字の範囲はスライス表記 [開始番号: 終了番号 -1] で取り出すことができます．たとえば，

文字列 'He' を取り出すには，s[0:2] とします．

●例題８－１

文字列 s = 'Hello World' から'World'を取り出してください．



In [ ]:
s = 'Hello World'
s[6:11]


スライス表記の開始番号と終了番号は省略できます．省略すると開始番号は 0 ，終了番号は文字列の長さ（これは len(文字列) で計算できます）になります．

また後ろから番号付けも可能で，最後の文字列の番号は -1，その前の文字列は -2 となります．

文字列は加算演算子 $+$ で結合することができます．たとえば，

'Hello' + 'World' は 'HelloWorld' になります（空白は挿入されません）．

また乗算演算子 $*$ も使うことができます．たとえば， $3*$ 'Hello' とすると 'Hello' +'Hello' +'Hello'  と同じ結果が得られます．

●例題８－２

5回 "I'm sorry!" と出力してみましょう．（文字列の中に'記号を入れたいときには，" "で囲みます．）

In [ ]:
print( 5 *"I'm sorry!")

for文を用いた反復はrange関数によって生成された数列だけでなく，文字列に対しても使えます．

たとえば，

```python
for x in 'Hello': 
   print( x )
``` 

と書くと，H, e, l, l, o が順番に変数 x に代入されて，print関数で出力されます．

●例題８－３

文字列 s='Kitty' の各文字の後ろに '!' を入れた文字列 x を出力しなさい．


In [ ]:
for x in 'Hello': 
    print(x)

In [ ]:
s = 'Kitty'
x = ''
for i in s:
    x = x + i + '!'
print(x)

●問題８－１

文字列を入力し，最初と最後が同じ文字なら 'Same!' と出力するプログラムを作りなさい．

In [ ]:
st = input("文字列を入力")
a = len(st)
if st[0] == st[a - 1]:
    print("Same!")
else:
    pass

●問題８－２

文字列を入力し，その中に 'a' が含まれているときに 'In!' と出力するプログラムを作りなさい．
（実は  *in* という演算子があるので「 'a' in 文字列」で簡単に 'a'を含むかどうかの判定はできるのですが，ここでは反復を用いて作って下さい．） 


In [ ]:
x = input()
a = len(x)
b = 0
for i in range(0, a):
    if x[i] == "a":
        b = b + 1
    else:
        pass
if b >= 1:
    print("In!")
else:
    pass

In [ ]:
x = input()
print("a" in x)

●問題８－３

文字列を引数とすると，各文字の間と最後に空白を入れた文字列を返値として返する関数を作って下さい．
さらに，文字列を入力し，その関数を呼び出して空白を入れた文字を出力するプログラムを書いて下さい．


In [ ]:
def space(x):
    y = ""
    for i in x:
        y = y + i + " "
    return y

x = input()
print(space(x))

# 9. リスト(list)

Pythonではオブジェクトを保管するための幾つかの便利な機能が（最初から）準備されています．

ここではその代表であるリストの使い方を説明します．

文字列が文字を箱に順番にしまっていたのと同じように，リストは何でも（Pythonではオブジェクトといいます）箱にしまっておくことができます．

リストは[ ]の中に *,*（カンマ）で区切ってものをしまいます．

```python
L = [ 1, 2, 3, 4]               # range(1,5)と同じように数列を保管
L = ['H', 'e', 'l', 'l', 'o']   # 文字列 'Hello' 同じようにう文字を保管
L = [ 'AAA', 235, ['BBB',1.3] ] # 文字列，整数，リストを保管

```

リストに対しても文字列と同じように，足し算，乗算，添え字（インデックス）による抽出や切り出し，for文による反復を行うことができます．

リストは文字列と違って可変(mutable)です（文字列は不変(immutable)です）．つまり中身を変えることができます．


●例題９－１

リスト L=[6, 5, 4, 3] の3番目の数字（Pythonでは0番から始まるので添え字2の数字）を'Hello'に変えて下さい．


In [ ]:
L=[6, 5, 4, 3] 
L[2] ='Hello'
print(L)

●例題９－２

大きさが100の変数のリストaを用意し，0番目に0，1番目に10，2番目に20，…と代入し，最後にそれらを表示する，というプログラムを作りましょう．

aの0番目から9番目までに代入するには，iが0から9まで１ずつ大きくなるfor ループを作り，a[i]に順々に数値を代入すればよいですね．

a[0]には0，a[1]には10，a[2]には20，．．．と代入したいので，a[i]に $i*10$ を代入します．

たとえば，以下のプログラムは，最初に0を入れたリストを10倍することによって，[0,0,0,0,0,0,0,0,0,0]というリストを準備をし，
次のforループで数字の代入を行います．

In [ ]:
a = 10 * [0]
for i in range(10):
    a[i] = i * 10
a

リストはいろいろな *メソッド* を適用できます．ここで，メソッドというのはリストに適用される関数のことです．

リストの後に *.* （ピリオド）を打った後に，「メソッド名(引数)」と書くと，関数のように出力が返ってきます．

たとえば，リストの最後に要素を追加するための append メソッドがあります．L.append( 'a' ) と書くと，リスト L の最後に文字列 'a' が追加されます．


上の例題９．２は，最初に空のリストを準備して，1つずつ数字をappendメソッドで追加していく方法でも解くことができます．


In [ ]:
a=[]
for i in range(10):
    a.append(i*10)
a

例題９．２では，[0,0,...,0] のリストを準備するために $10 *[0]$ と乗算を使いました．

これは以下の *リスト内包表記* と同じですが，リスト内包表記の方が一般的なので推奨される書き方です．

```python
a=[0 for i in range(10)]
```

これは for i in range(10) で iを0から9まで変えた反復を行い，反復ごとに 0 を入れたリストを生成する，という意味です．

0 のかわりに $i*10$ と書けば問題は 1行で解けます．

In [ ]:
a=[i*10 for i in range(10)]
a


リストの要素を小さい順に並べ替えたいときには，sort()メソッドを使います．
たとえば，
```python
L =[ 6, 8, 3, 1] 
L.sort()
``` 
とすると，Lは [1, 3, 6, 8] と小さい順に並べ変わります．

逆順（大きい順）に並べ替えたいときには，sort()メソッドを使った後に，reverse()メソッドを使います．

●例題９－３

最初の行にimport randomと宣言した上で，x=random.randint(a,b)と書くことで，xにaからbまでのランダムな整数を代入してくれます．

ランダムに生成した1から100までの整数から成る長さ 10 のリストを作り，それを小さい順，大きい順に並べ替えたリストを出力しなさい．


In [ ]:
import random
L = []
for i in range(10):
    L.append( random.randint(1,100))
print('元のリスト ', L)
L.sort()
print( '小さい順 ', L)
L.reverse()
print('大きい順 ', L)

●問題９－１

キーボードから10個の数を入力し，それらを入力された順番と逆の順番で表示するプログラムを作りなさい．

In [ ]:
a = []
for i in range(10):
    a.append(int(input()))
b = [0] * 10
for i in range(10):
    b[i] = a[9 - i]
print(b)

●問題９－２

まず，長さ10に全部 0 が入ったリストを作りなさい．

次に，0から9までの整数 a1,a2を入力し，リストxのa1番目からa2番目までを1大きくする，という作業を繰り返すプログラムを作りなさい．

このとき，毎回リストの0番目から9番目までを表示するようにしなさい．

（ヒント： iをa1からa2までの１ずつ大きくするforループを使い，リストのi番目を1大きくするようにしましょう．）

In [ ]:
a = [0] * 10
while True:
    a1 = int(input("a1入力"))
    a2 = int(input("a2入力"))
    for i in range(a1, a2 + 1):
        a[i] = a[i] + 1
    print(a)

●問題９－３

リストと整数iを受け取り，リストのi番目の値を１大きくする関数を作りなさい．また，この関数を試すプログラムを作りなさい．


In [4]:
def lst(x, i):
    x[i] = x[i] + 1
    return x

x = [4, 7, 8, 1, 3, 7]
i = 1
print(lst(x, i))

[4, 8, 8, 1, 3, 7]


In [2]:
x = int(input("listの次元"))
l = []
for i in range(x):
    a = int(input())
    l.append(a)
print(l)

listの次元3
0
1
2
[0, 1, 2]


●問題９－４

“Taro”, “Jiro”,”Saburo”,”Shiro”,”Goro” がこの順番で座っている．
この4人を名前を入れたリスト L は

```
L=[“Taro”, “Jiro”,”Saburo”, ”Shiro”, ”Goro”]
```
となる．これを参考にして，自分の座っている列にいる友人の名前のリストを作成せよ．

In [ ]:
L = ["A", "B", "C", "D", "E"]

●問題９－５

上で作成したリストに対して，最初に座っている人と最後に座っている人を表示せよ．

また，名前の昇順のリストを作成せよ．


In [11]:
L = ["e", "d", "c", "b", "a"]
print(f"最初に座っている人={L[0]}")
print(f"最後に座っている人={L[-1]}")

L.sort()
print(L)

最初に座っている人=e
最後に座っている人=a
['a', 'b', 'c', 'd', 'e']


●問題９－６

100個の[1,10000]の一様整数乱数のリストを作成し，次にそれらの和と平均を求めよ．

（ヒント：リスト L の和は関数 sum(L) で，要素数は関数 len(L) で求めることができます．）

また，平均に最も近い数を求め，それが何番目の数字（最初の数字を0番目とする）なのかを計算するプログラムを作成せよ．

In [54]:
import random
l = []
for i in range(100):
    l.append(random.randint(1,10000))
a = sum(l)
b = len(l)
c = a / b
print(f"リストの和={a}、リストの平均={c}")

m = 10000
n = 0
for i in range(100):
    m = min(m, abs(l[i] - c))
    if m == abs(l[i] - c):
        n = i

print(f"平均に最も近い数 = {l[n]}")
print(f"{n}番目")

リストの和=499598、リストの平均=4995.98
平均に最も近い数 = 4938
17番目


# 10. タプル(tuple)

タプルはリストとほとんど同じものですが，リストと違って中身が変更できない不変型(immutable)です．
文字列と同様に（不変型なので）以下で説明する辞書のキーとして使うことができます．

タプルを生成するには () の中に *,* で区切って入力します．

```python
  T =(6,4,7,'Hello')
 ```
 
タプルはリストと同様に添え字（インデックス）や切り出しができますが，中身を変えることができません．

例題１０－１

上のタプルTの4番目の要素（Pythonの添え字は0から始まるので添え字3の要素）を出力せよ．

また，4番目の要素を数字の6に変更してみよ．（エラーが発生するはずです．）


In [35]:
T =(6,4,7,'Hello')
print( T[3])
T[3]=6

Hello


TypeError: 'tuple' object does not support item assignment

例題１０－２

a = 'Hello' と b='World' の2つの変数の中身を1行で取り替えよ．

（aとbから成るタプル (a,b) に (b,a)を代入すると1行で取り替えることができます．

また，() は省略して書くこともできます．

ただし1つの要素から成るタプルの場合には省略できず，*(a,)* とカンマ *,* を入れて書きます． ）


In [36]:
a = 'Hello'
b = 'World'
a,b = b, a
print(a,b)

World Hello


問題１０－１

変数 a,b,c に代入された文字列を変数 x,y,z に代入せよ（ただし1行で）．
```python
a = 'Mickey' 
b = 'Minny'
c = 'Kitty' 
```

In [37]:
a = 'Mickey' 
b = 'Minny'
c = 'Kitty' 
x, y, z = a, b, c
print(x, y, z)

Mickey Minny Kitty


変数 a,b,c に代入された文字列を a=>b, b=>c, c=>a になるように交換せよ（ただし1行で）．
```python
a = 'Mickey' 
b = 'Minny'
c = 'Kitty' 
```

In [39]:
a = 'Mickey' 
b = 'Minny'
c = 'Kitty' 
a, b, c = c, a, b
print(a, b, c)

Kitty Mickey Minny


# 11. 辞書(dictionary)

リストの他に辞書という便利な機能もあります．辞書は，*キー* を入れると *値*（あたい）が返ってくる写像(マップ）のようなものです．
順番は任意なのでマップ型と呼ばれます．

これは通常の辞書で，"Hello" をひくと「こんにちわ」が書いてあるのと似ています．

辞書は {} をの中に「キー：値」を *,* （カンマ）で区切って入れます．英語を日本語に変換する辞書を作ってみましょう．

```python
D ={ 'Hello' : 'こんにちわ',  'Good Morning': 'おはよう', 'Good Night': 'おやすみ' }

D['Good Morning'] 
```

最初の行で辞書 D を作り，次の行で辞書Dからキー'Good Morning'を引きます．これは対応する値 'おはよう' を返します．

---
3人の女の子の身長を入れた辞書は以下のようになります．

```python
D={ 'Mary': 126, 'Jane': 156, 'Sara': 170} 
```

キーのリストは D.keys() で値のリストは D.values() で得ることができます．

キーになれるのは不変型(immutable)だけです．したがって，リストや辞書などの可変型(mutable)のオブジェクトはキーにはなれません．

キーには数字，文字列，タプルなどを使ってください．

値は任意のオブジェクトを入れることができます．リストや辞書でも大丈夫です．

たとえば，値に身長と体重のリストを入れた辞書は以下のようになります．

```python
D={ 'Mary': [126,30], 'Jane': [156,55], 'Sara': [170,60]} 
```

辞書 D の中にキーがあるかどうかは「キー in D」と*in* 演算子を用いて判定することができます．

たとえば，上で作った身長の辞書Dに対して
```
  'Kitty' in D 
```
は False を返し，
```
  'Mary' in D 
```
は True を返します．


辞書に新しい要素を追加したり，値を変更したいときは，単に「D[キー]=値」と書きます．
たとえば，身長が 10cmの'Kitty'を追加するには
```python
 D['Kitty'] =10 
```
と書き，'Jane'の身長を180に変更したいときには，
```python
 D['Jane'] =180 
```
と書きます．


辞書 D に対してもfor文を使うことができます．構文は for x in D: です．このとき x に代入されるのはキーですが，順番は入力した順ではありません．
（リストや文字列は順序をもつ型（順序型）なので順番が保たれますが，辞書はマップ（写像）型なので順番はばらばらになります．）

●例題１１－１

上で作成した身長の辞書から身長が160cm以上の人の名前を出力せよ．


In [55]:
D={ 'Mary': 126, 'Jane': 156, 'Sara': 170} 
for i in D:
    if D[i] >=160:
        print (i)

Sara


●例題１１－２

上で作成した身長の辞書を，小さい順に並べたときの名前を出力せよ．

まず，身長と名前のタプルを入れたリストをfor文を用いて作ります(この時点では並び順は任意です）．

次にそれを sort()メソッドで小さい順に並べ替えます（並べ替えはタプルの前の要素（この場合には身長）が優先されます）．

並べ替えた後のリストをfor文を用いて反復し，名前を順に出力します．


In [73]:
D = {'Sara': 170, 'Jane': 156, 'Mary': 126} 
L = []   #空のリストを準備しておく
for i in D:
    L.append( (D[i], i) ) #（身長，名前）のタプルをリストに追加
print('並べる前のリスト' ,L)
L.sort()
print('身長順のリスト ' ,L)
for (height,name) in L:
    print(name)

並べる前のリスト [(170, 'Sara'), (156, 'Jane'), (126, 'Mary')]
身長順のリスト  [(126, 'Mary'), (156, 'Jane'), (170, 'Sara')]
Mary
Jane
Sara


In [69]:
D = {'Mary': 126, 'Jane': 156, 'Sara': 170}   #'Mary'…キー(key)と'126'…値(value)
print(type(D))
print(D['Mary'])
D['Mary'] = 130
D['Un'] = 120
print(D)

print('Mary' in D)
print(len(D))
del D['Un']
print(D)

<class 'dict'>
126
{'Mary': 130, 'Jane': 156, 'Sara': 170, 'Un': 120}
True
4
{'Mary': 130, 'Jane': 156, 'Sara': 170}


●問題１１－１

名前をキー，体重を値とした辞書 D を作成せよ．ただし，友人は3人以上記入し，
名前には友人の名前を文字列として，体重には友人の理想体重を整数値として入力せよ．

In [74]:
D = {'An': 60, 'Moa': 70, 'Un': 80}
print(D)

{'An': 60, 'Moa': 70, 'Un': 80}


●問題１１－２

起動すると，友人の名前の入力を求め，上で作成した辞書Dから体重を調べて画面上に表示するプログラムを作成せよ．

In [79]:
D = {'An': 60, 'Moa': 70, 'Un': 80}

x = input("名前の入力")
print(f"{x}の体重:{D[x]}")

名前の入力An
Anの体重:60


●問題１１－３

上のプログラムを，何度も友人の名前の入力を求めるように変更せよ．ただし，辞書にない名前を入力すると終了するものとする．

In [80]:
D = {'An': 60, 'Moa': 70, 'Un': 80}

while True:
    x = input("名前の入力")
    if x not in D:
        break
    print(f"{x}の体重:{D[x]}")

名前の入力An
Anの体重:60
名前の入力Moa
Moaの体重:70
名前の入力Un
Unの体重:80
名前の入力a


●問題１１－４

空の辞書Dを準備し，その後で名前と体重の入力を何度もユーザーに要求して，
名前をキー，体重を値としてDに保管していくプログラムを作成せよ．
ただし，ユーザー名に空白 ”” を入力すると，画面に作成した辞書Dを表示してから，終了するものとする．


In [3]:
D = {}
while True:
    x = input("名前入力")
    if x == " ":
        print(f"D = {D}")
        break
    else:
        y = int(input("体重入力"))
        D[x] = y

名前入力Un
体重入力80
名前入力Moa
体重入力70
名前入力An
体重入力60
名前入力 
D = {'Un': 80, 'Moa': 70, 'An': 60}


●問題１１－５

友人の名前と体重を保管した辞書Dを準備し，その後で，友人の名前と体重の入力を求め，名前が一致する友人がいたならば，
体重を新しい値に変更し，いなければ辞書に追加するプログラムを作成せよ．
ただし，ユーザー名に空白 ”” を入力すると，画面に作成した辞書Dを表示してから，終了するものとする．

In [9]:
D = {'Mary': 30, 'Jane': 40, 'Sara': 50, 'An': 60, 'Moa': 70, 'Un': 80}
while True:
    x = input("名前入力")
    if x == " ":
        print(f"D = {D}")
        break
    else:
        y = int(input("体重入力"))
        D[x] = y

名前入力Mary
体重入力1000
名前入力Miffy
体重入力50
名前入力 
D = {'Mary': 1000, 'Jane': 40, 'Sara': 50, 'An': 60, 'Moa': 70, 'Un': 80, 'Miffy': 50}


●問題１１－６

問題１１－１で作成した友人の名前と体重を保管した辞書Dの名前を体重を名前のアルファベット順に画面に出力するプログラムを作成せよ．

In [25]:
D = {'Mary': 30, 'Jane': 40, 'Sara': 50, 'An': 60, 'Moa': 70, 'Un': 80}
D1 = sorted(D.items())
print(D1)
D2 = sorted(D.items(), key = lambda x:x[1])    #x[0]とするとアルファベットの並び順になる
print(D2)
D3 = sorted(D.items(), key = lambda x:x[0])
print(D3)

[('An', 60), ('Jane', 40), ('Mary', 30), ('Moa', 70), ('Sara', 50), ('Un', 80)]
[('Mary', 30), ('Jane', 40), ('Sara', 50), ('An', 60), ('Moa', 70), ('Un', 80)]
[('An', 60), ('Jane', 40), ('Mary', 30), ('Moa', 70), ('Sara', 50), ('Un', 80)]


In [23]:
fruits = {'Orrange':150, 'Apple':100, 'Grape':400, 'Banana':80}
fruits = sorted(fruits.items())
print(fruits)
fruits = dict((x, y) for x, y in fruits)
print(fruits)

[('Apple', 100), ('Banana', 80), ('Grape', 400), ('Orrange', 150)]
{'Apple': 100, 'Banana': 80, 'Grape': 400, 'Orrange': 150}


●問題１１－７

好きな都道府県を5つ選択した後，都道府県名をキーとし，その名産品（1つ以上入力する）のリストを値とした辞書Dを作成せよ．

また，作成した辞書に対して，名産品の多い順に並べるプログラムを作成せよ．

(ヒント：リストLの長さ(要素数）は関数len(L)で計算できます．）

# 12. クラス

Pythonでは，数やリストや辞書などはすべてオブジェクトと呼ばれます．これらのオブジェクトは元から準備されたものなので組み込みオブジェクトと呼ばれます．

Pythonでは，自分で新しいオブジェクトを作ることができます．それがクラス(class)です．

構文は関数と似ています．関数を定義するキーワード def の代わりに class と書きます．

```python
class クラス名():
    クラス本体
```

クラスとは，オブジェクトの属性やオブジェクトに対する操作（これをメソッドと呼びます）をまとめたものです．

たとえば，ピカチュウというクラスは，「ほっぺた」や「しっぽ」などの属性と，「しっぽを振る」とか「リンゴを食べる」といったメソッドをもちます．

例として「点」を表すクラス Point を作ってみましょう．

```python
class Point():
    pass

p1=Point()
```

pass は何もしないことを表します（何も入れないとエラーします）．クラス名は Point で，```p1=Point()```でp1というクラスの実体（インスタンスと言います）を生成しています．

p1 という点を表すインスタンスには，x座標とy座標という**属性** (attribute)をもたせることができます．

属性はインスタンスの後ろに **.** を書いた後に記述します．

```python
p1.x = 100
p1.y = 200
```
これで点p1にx,y座標の属性100,200をもたせることに成功しました．

次に，点を移動させるとか，座標をまとめて出力するといった**メソッド**を記述しましょう．

メソッドは，クラス定義の**中**で関数と同様に def と書くことで定義できます．

ただし，メソッドの最初の引数には必ず自分自身を表す**self**を入れます．クラス内でx属性にアクセスするときには，**self.x**と記述します．

●例題１２－１

点を表すクラス Point を作成し，x座標とy座標を表す属性に100,200を代入してください．

次に，座標の出力を行うメソッドprintMeを用いて座標の出力を行ってください．

また，座標の移動を行うメソッドmoveを定義し，座標を右に10，上に20移動させてから座標の出力を行ってください．


In [1]:
class Point():
    def printMe(self):
        print( "<", self.x, ",", self.y, ">" )
    def move(self,right,up):
        self.x += right
        self.y += up

p1 = Point()
p1.x = 100
p1.y = 200

p1.printMe()

p1.move(10, 20)
p1.printMe()

< 100 , 200 >
< 110 , 220 >


In [16]:
class Point():
    def printMe(self):
        print( "<", self.x, ",", self.y, ">" )
    def move(self,right,up):
        self.x += right
        self.y += up

p1 = Point()
p1.x = 100
p1.y = 200

p1.printMe()

p1.move(10, 20)
p1.printMe()

p1.move(-60, -90)
p1.printMe()

< 100 , 200 >
< 110 , 220 >
< 50 , 130 >


クラスには```__``` (アンダースコア2つ）で挟んだ特殊メソッドを定義することができます．

たとえば文字列を返す**__str__**を定義すると，クラスのインスタンスをprintしたときに，その文字列が表示されるようになります．

クラスの初期化を行うメソッド**__init__**を定義すると，インスタンスを最初に生成したときに「自動的に」呼ばれ，インスタンスの初期化を行うことができます．

たとえば，点のインスタンスを生成するときに「必ず」x,y座標を与えて，x,y属性に代入するには，

```python
    def __init__(self, x, y):
        self.x=x
        self.y=y
```
と記述します（最初の引数に**self**を入れるのを忘れないようにしてください）．

ただし，点クラスのインスタンスを生成するときには，**self**は書かずに，以下のように記述します．

```python
p1=Point(100,200)
```

●例題１２－２

初期化をするためのメソッド```__init__```を用いて点を表すクラス Point を作成し，x座標とy座標を表す属性に100,200を代入してください．

また，文字列を返す```__str__```を用いて座標の出力を行ってください．

さらに，他の点インスタンスの座標を加える特殊メソッド```__add__```（これは + 演算子を適用したときに呼ばれます）を
使っての座標 (10，20) をもつ点インスタンスとの足し算をしなさい．
        

In [3]:
class Point:
    def __init__(self, x, y):
        self.x = x
        self.y = y
        
    def __str__(self):
        return "<"+ str(self.x) + ", " + str(self.y) +">"
    
    def __add__(self, other):
        return Point( self.x+other.x, self.y+other.y )
    
p1 = Point(100, 200)
print( p1 )
p2 = Point(10, 20)
print( p1+p2 )

<100, 200>
<110, 220>


●問題１２－１

身長と体重の2つの属性をもつHumanクラスを作成してください．

その後に，「体重/身長の2乗」で定義されるボディマス指数を計算するメソッド bmi を追加してください．
（ただし，体重はキログラム，身長はメートルで入れます．）

次に，以下の表に基づいて，痩せ型から肥満(4度）までのいずれに属するかを文字列として返すメソッド am_i_fat を追加してください． 

                     BMI値
    痩せ型     	18.5未満
    普通体重	   18.5以上、25未満
    肥満（1度）	25以上、30未満
    肥満（2度）	30以上、35未満
    肥満（3度）	35以上、40未満
    肥満（4度）	40以上
    
自分の身長と体重（公表するのが嫌な場合には，友人の身長と体重）を属性としたHumanのインスタンス me を作成して，am_i_fatを呼んでみましょう．



In [15]:
class Human():
    def Me(self):
        print("<", self.height, ",", self.weight, ">")
        
    
    def am_I_fat(self):
        BMI = ( (self.weight) / (self.height) ** 2 )
        if BMI < 18.5:
            print("痩せ型")
        elif 18.5 <= BMI < 25:
            print("普通体重")
        elif 25 <= BMI < 30:
            print("肥満（1度）")
        elif 30 <= BMI < 35:
            print("肥満（2度）")
        elif 35 <= BMI < 40:
            print("肥満（3度）")
        else:
            print("肥満（4度）")

p1 = Human()
p1.height = float(input("身長を入力[m]"))
p1.weight = float(input("体重を入力[kg]"))
p1.Me()

p1.am_I_fat()

身長を入力[m]1.75
体重を入力[kg]70
< 1.75 , 70.0 >
普通体重


●問題１２－２

例題で作った点を表すクラスを参考にして，ベクトルを表すクラスを定義し，2つのベクトルの内積を乗算 $*$ で行うようなメソッド
```__mult__```を書いてみましょう．

In [ ]:
x=3
y=x 
print(y is x)

y=3
print(y is x)

L=[1,2,3]
M=[1,2,3]

print(M == L)
print(M is L)

M=L

print(M is L)

---
## (補足)オブジェクトの型について

### A.1.1 数と論理型

type関数を用いて，以下の出力結果を得なさい．

(1). float

(2). int

(3). complex 

(4). bool

### A.1.2 文字列

1. 文字列 s='PythonVideoCheckProblem' から，スライス表記を用いて 'onVide' を抜き出しなさい．

2. 文字列s='PythonVideoCheckProblem' をすべて大文字にしなさい．

3. 以下の文字列，並びに format を用いて，'Python,Video,Check,Problem' を作成しなさい．なお，単語の間には ',' を挟むこと．
```python
s1 = 'Problem'，
s2 = 'Python'，
s3 = 'Check'，
s4 = 'Video'
```

4. 引数として与えた文字列が全て小文字なら True．そうでないなら Falseを返す関数を作りなさい．
例として，文字列 s='PyThOn' を関数の引数として与えて判定しなさい．

5. 小文字として与えた文字列の先頭が母音（'a','i','u','e','o'）である場合には，'an 'を，そうでない場合には 'a 'を加えた文字列を返す関数を作れ．
例として，'apple'と'pen'を引数として与えた関数の返値を出力せよ．


### A.1.3 リスト

1. リスト L = [ 'Marvelous', 'Fantastic', 'Amazing!!' ] 内の文字列を，s='Python!! ' を区切り文字として結合し，表示しなさい．

2.  文字列s = 'Payatahaoan' のうち，'a' を区切り文字として文字列s を分割したリストで表示しなさい．

3. リスト L = [1,2,3,4,5] が与えられている．

(1). ２番目の要素を削除しなさい．

(2). L の2番目の添え字に 300 を挿入しなさい．

(3). L を逆順にしなさい．

(4). スライス表記を用いて，2番目に [10,20] を代入しなさい．

(5). メソッド append を用いて，100 を追加しなさい．


### 問題

1. 文字列 s='May I have a large container of coffee'を空白を区切り文字として分割し，分割した文字列の長さを順に表示せよ．何になるか？

2. 文字列 s='How I want a drink, alcoholic of course, after the heavy lectures involving quantum mechanics, and if the lectures were boring or tiring, then any odd thinking was on quartic equations again' を空白を区切り文字として分割し，分割した文字列の長さを順に表示せよ．何になるか？

### A.1.5 辞書

1. 辞書 D = { 'Mary':126, 'Jane':156, 'Sara': 170} が与えられている．
 1. キー'Mary'の値，126を180に変更しなさい．
 2. メソッドを用いて，キーのリストを表示しなさい．
 3. メソッドを用いて，値のリストを表示しなさい．

2. 与えられた正数 $n$ に対して，$n$ 行 $n$ 列のランダムな行列で，各要素が[1,10]の一様整数乱数の行列を，行と列のタプルをキーとし，要素を値とした辞書を用いて作成せよ．（タプルは変更できないという一点を除いてリストと同じように扱うことができる．タプルは不変（中身の変更が不可である）オブジェクトなので，辞書のキーとして使うことができる．）

3. 同じ大きさの行列をもう1つ作成し，行列の積を求めよ．（forループを用いて計算しても良いが，Python 3.5以降だと @ で掛け算ができる．）

4. 名前をキー，体重を値とした辞書 D を作成せよ．ただし，友人は3人以上記入し，名前には友人の名前を文字列として，体重には友人の理想体重を整数値として入力せよ．

5. 友人の名前の入力を求め，上で作成した辞書Dから体重を調べて画面上に表示するプログラムを作成せよ．

6. 上のプログラムを，何度も友人の名前の入力を求めるように変更せよ．ただし，辞書にない名前を入力すると終了するものとする．

7. 空の辞書Dを準備し，その後で名前と体重の入力を何度もユーザーに要求して，名前をキー，体重を値としてDに保管していくプログラムを作成せよ．ただし，ユーザー名に空白 ''（空の文字列）を入力すると，画面に作成した辞書Dを表示してから，終了するものとする．

8. 友人の名前と体重を保管した辞書Dを準備し，その後で，友人の名前と体重の入力を求め，名前が一致する友人がいたならば，体重を新しい値に変更し，いなければ辞書に追加するプログラムを作成せよ．ただし，ユーザー名に空白 ”” を入力すると，画面に作成した辞書Dを表示してから，終了するものとする．

9. 上で作成した友人の名前と体重を保管した辞書Dに対して，友人の名前を１行ずつすべて画面に印刷するプログラムを作成せよ．

10. 上で作成した友人の名前と体重を保管した辞書Dに対して，保管されている体重を１行ずつすべて画面に印刷するプログラムを作成せよ．

11. 上で作成した友人の名前と体重を保管した辞書Dの名前を体重を名前のアルファベット順に画面に印刷するプログラムを作成せよ．

12. 上で作成した友人の名前と体重を保管した辞書Dの名前を体重を体重の軽い順に画面に印刷するプログラムを作成せよ．

13. 都道府県名をキーとし，その名産品（1つ以上入力する）をリストとした辞書Dを作成せよ．都道府県は自分の出身地を入れて3つ以上準備せよ．
次に，作成した辞書に対して，名産品の多い順に並べるプログラムを作成せよ．


### A.1.6 集合

集合S = { 1,2,3,4,5 } と集合T = { 2,4,6 } が与えられている．

(1). SとTの和集合を表示しなさい．

(2). SとTの差集合を表示しなさい．

(3). SとTの共通部分を表示しなさい．

(4). 集合Sの要素 2 を削除し，もう一度SとTの共通部分を表示しなさい．

(5). “banana”の各文字なら成る集合と，”apple”の各文字から成る集合の和集合と共通部分を計算するプログラムを示せ．

### A.2 演算子

Python3系では通常， 5 / 3 の出力結果は，1.6666666……となってしまう．これを，整数部分のみ出力を得たい場合はどのような演算子を用いるべきか．

### A.4 関数

1. 引数を2つ与え、それらの和を出力する関数 printSum( x, y ) を作成し，その関数に規定値を与え，printSum() と引数なしで実行した場合，出力結果 10を得られるように作成しなさい．

2. 無名関数 lambda を用いて，与えた数の2乗が返るような関数を作成し，表示しなさい．

3. 文字列を合体し，3回繰り返したものを返す関数を作成しなさい．


### 再帰 例題

「1円玉1枚，10円玉1枚，100円玉1枚の組合せでできる8通りの金額を全て表示する」というプログラムを，再帰呼び出しを使って作りましょう．

まず第1ステップとして，1円，10円，100円の全て使った組合せの金額を，「再帰呼び出し」を使って計算するプログラムを作りましょう（単に足せば良いのですが，再帰を使って計算してみます）．

計算を行うための関数payの引数は，使うかどうか試したコインの数(iという名前にしました），合計を記録する補助変数(totalという名前にしました)，とリスト coin = [1, 10, 100] です．再帰の終了条件は，すべてのコインを使い終わったら（つまりiが3になったら）合計金額を出力することにします．

合計金額は，コイン数を1増やし（iを1増やし），合計金額totalを coin[i] だけ増やして，payを再帰的に呼ぶことによって計算されます．

In [ ]:
def pay(i,total,coin):
    if i==len(coin):
        print('total=',total)
    else:
        pay(i+1, total+coin[i], coin)
    
coin = [1,10, 100]
pay(0, 0, coin) 

次に，全ての組合せを使うプログラムを作ります．全ての組合せの金額を表示するには，各硬貨について，「その硬貨を使う場合」と「その硬貨を使わない場合」の場合分けが必要です．そこで，それぞれの再帰呼び出しが，その硬貨を使う組合せ，使わない組合せに対応します．

In [ ]:
def pay(i,total,coin):
    if i==len(coin):
        print('total=',total)
    else:
        pay(i+1, total, coin)         #コイン iを使わない場合
        pay(i+1, total+coin[i], coin) #コイン iを使う場合
    
coin = [1,10, 100]
pay(0, 0, coin) 

### 問題

1. 2つの数mとn(ただし，m>n )が与えられたとき，mとnの最大公約数は，nとm%n (mをnで割った余り)の最大公約数になります．これを利用して，入力した2つの整数の最大公約数を求めるプログラムを，再帰呼び出しを使って作りなさい．

2. 回文(palindrome)とは前から読んでも後ろから読んでも同じ文字列を指す．引数として与えた文字列が回文であるか否かを判定する関数を，再帰を用いて記述せよ．例として 'たけやぶやけた' と 'たけたぶやけてない' を入れて試せ．

3. 5つの正数を入れたリスト a と正数 bを入力して，a[0],...,a[4]の合計がちょうどbになる組合せがいくつあるかを表示するプログラムを再帰呼び出しを用いて作りなさい．


### A.5 内包表記

1. $0$ から $10$ までの整数のリストを内包表記で表示しなさい．

2. $0$ から $5$ までの整数 $x$ に対して， $(x, x^2, 2^x)$ からなるタプルのリストを内包表記で作成しなさい．

3. ジェネレータ内包表記とsum関数を用いて，$0$ から $10$ までの整数でかつ，奇数のものの合計を表しなさい．

4. $0$ から $5$ までの整数に対して，整数がキー，値がキーの2乗になるような辞書を，内包表記を用いて表しなさい．

### A.8 モジュール

1. mathモジュールをインポートし，

 1. $2$ の平方根を表示しなさい．

 2. $6$ の階乗を表示しなさい．(階乗を表示するには，factorial() を用います)

2. random モジュールから，randint関数, shuffle関数 のみをインポートし、

 1. $-5$ 以上，$3$ 以下の一様ランダムな整数を発生させなさい．

 2. リスト $L = [1,2,3,4]$ の順序をランダムに混ぜなさい．

3.  itertoolsモジュールをインポートし，

 1. range(5)内の要素の長さ $2$ の（繰り返しを許さない）順列を生成し，リスト型で表示しなさい．
 2. range(5)内の要素の長さ $2$ の（繰り返しを許さない）組み合わせを順に生成し，リスト型で表示しなさい．

### randomモジュール

1. 各期の需要が平均100，標準偏差10の切断正規分布（負の場合には0とした正規分布）のとき，100期分のシミュレーションを行い，需要が110を超える期数を調べるプログラムを作成せよ．（ヒント：正規分布にしたがう擬似乱数はrandomモジュールのgauss関数（もしくはnormalvariate関数）で生成できる．）

2. あるスーパーで販売しているある商品の在庫を管理したい．
いま，$t-1$日の在庫量の営業後に在庫を調べて，発注量 $x[t]$ を決める．発注した商品は翌日の朝に届き，
それによって $t$ 日の営業後の在庫 $I[t]$ は，以下のように決まるものとする．

$$
I[t] = I[t-1]+x[t] –D[t]
$$

ここで $D[t]$ は$t$日の需要量であり，需要が平均100，標準偏差10の切断正規分布（負の場合には0とした正規分布）とする．
また，最初の日( $t=0$)の営業後の在庫は0であると仮定する．

以下の2つの在庫管理方策のプログラムを作成せよ．

* (s,S)方策：在庫量 $I[t-1]$ が $s$ 未満であるなら，翌朝の在庫が $S$ になるように発注する．

* (Q,R)方策：在庫量 $I[t-1]$ が $R$ 未満であるなら，$Q$ だけ発注する．

 100日のシミュレーションの後で，発注費用，在庫費用，品切れ費用の合計を計算するプログラムを作成せよ．ただし，毎日の営業後の在庫量が0以上の場合には1個あたり10円の在庫費用がかかり，在庫量が負の場合には，1個あたり100円の品切れ費用がかかるものとする．（負の在庫は顧客が待っている状態を表し，翌日の朝に商品を引き取ると仮定する．これをバックオーダーと呼ぶ．）また，発注を行う度に（発注量が正のときに）発注費用1000円がかかるものとする．また，作成したプログラムを用いて，各方策のパラメータはどのように設定すれば総費用が最小になるか考察せよ．

In [ ]:
import random
move = 0
dont_move =0
for iter in range(10000):
    win = random.choice([1,2,3])
    select = random.choice([1,2,3])
    remainning = {1,2,3} - {select}
    removed = random.choice(list(remainning))
    if select == win:
        dont_move +=1
    else:
        move+=1
print(move,dont_move)

In [ ]:
removed

In [ ]:
remainning